ansible-ipi-install/roles/installer/templates/install-config.j2

```diff
@@ -74,7 +74,7 @@ platform:
           address: redfish://{{ hostvars[host]['ipmi_address']|ipwrap }}/redfish/v1/Systems/1
           disableCertificateVerification: {{ disable_bmc_certificate_verification }}
 {% else %}
-          address: ipmi://{{ hostvars[host]['ipmi_address']|ipwrap }}:{{ hostvars[host]['ipmi_port']|default(623) }}
+          address: {{ bmc_type | default('ipmi') }}://{{ hostvars[host]['ipmi_address']|ipwrap }}:{{ hostvars[host]['ipmi_port']|default(623) }}
```

Comment:
So if I understand correctly you changed your inventory/hosts file to have irmc_address, irmc_user, irmc_password, etc... and then updated the template to use that instead of ipmi.

You mention power off hosts with ipmi will fail with my version.  How does it work in your version?  or does it just skip it since there isn't any ipmi_ variables?

Reply:
> You mention power off hosts with ipmi will fail with my version. How does it work in your version? or does it just skip it since there isn't any ipmi_ variables?

No, it wouldn't skip power off unless we specify `poweroff=false in inventory/hosts file.

Sorry I may not have been detailed enough before, let me restate.

Firstly, let's take a look at the difference between your and my version of [templates](https://github.com/openshift-kni/baremetal-deploy/blob/master/ansible-ipi-install/roles/installer/templates/install-config.j2#L77) and [power off tasks](https://github.com/openshift-kni/baremetal-deploy/blob/master/ansible-ipi-install/roles/installer/tasks/59_power_off_cluster_servers.yml#L15-L20) according to the hosts configuration
Your version, template:
```
address: irmc://127.0.0.1:443
```
Your version,  power off task:
```
- name: Power off hosts
      ipmi_power:
        name: 127.0.0.1
        ...
        port: 443
```

My version, template:
```
address: irmc://127.0.0.1:443
```
My version,  power off task:
```
- name: Power off hosts
      ipmi_power:
        name: 127.0.0.1
        ...
        port: 623
```
It can be seen that except the port number of power off tasks is different, everything else is the same. Because the port of irmc is 443, so I have to specify `ipmi_port=443`. The power off task also use ipmi_port 443, thus cause the failure in your version. The following is the host configuration written according to your version:
```
master-0 name=master-0 role=maste ipmi_user=<user> ipmi_password=<passwd> ipmi_address=<IP> ipmi_port=443 bmc_type=irmc provision_mac=<mac_address>
```

In my version, I replace all the ipmi with irmc in the template, so I add irmc_address, irmc_user, irmc_password and irmc_port to the inventory/hosts, besides I also add ipmi_address, ipmi_user, ipmi_passwd to the inventory/hosts for power off task. Because I didn't specify ipmi_port, so power off task will just use the default port 623, thus the power off task can run successfully. The following is the host configuration written according to my version:
```
master-0 name=master-0 role=master irmc_user=<user> irmc_password=<passwd> irmc_address=<IP> irmc_port=443 provision_mac=<mac_address> ipmi_user=<user> ipmi_password=<passwd> ipmi_address=<IP>
```

Reply:
If the power off works via the ansible IPMI module with the IPMI port then why doesn't install-config.yaml work with ipmi:// as well?  I really don't understand the mix of irmc and ipmi.  It's very confusing.

Reply:
> why doesn't install-config.yaml work with ipmi:// as well?

We also want to test the configuration of RAID and BIOS in IPI process, but `ipmi` does not support this, so I need to specify the bmc as `irmc`.
btw, from OCP4.10, IPI has support the configuration of RAID and BIOS for worker nodes.

Reply:
> I've pushed a new version which should handle your use case

Thanks, it works.

> One thing I'm not sure about though is the current version is using the ipmi user and password, Will that work for you?

For now, the user and passwd of our irmc and ipmi are same, so it works. But if they are different, the current template won't work for it.

Reply:
That was my question.  So they can be different? I'll update the template to be explicit.

Reply:
> So they can be different?

Sorry I made an unfounded assumption yesterday. They should be same, the user and password should belong to BMC, other than ipmi or irmc.

> I'll update the template to be explicit.

It looks that your latest change will cause missing user and password for dell and hp's redfish.

IMO, it‘s fine to go back to previous version.

Reply:
done.  please test and report here.

thanks

Reply:
Thanks, this works fine for us.